src/reactor.rs

```
@@ -0,0 +1,106 @@
+    },
+}
+
+/// Anything that wants to react on block generation should implement this.
+pub trait React {
+    /// Defines how different reactor will react on block observation.
+    ///cargo
+    /// # Arguments
+    ///
+    /// * `block` - The observed block.
+    /// * `event_loop` - The reactor's event loop to handle the tasks spawned.
+    fn react(&self, block: &Block, event_loop: &tokio_core::reactor::Handle);
+}
+
+impl React for Reactor {
+    /// Defines how different reactor will react on block observation.
+    ///
+    /// # Arguments
+    ///
+    /// * `block` - The observed block.
+    /// * `event_loop` - The reactor's event loop to handle the tasks spawned.
+    fn react(&self, block: &Block, event_loop: &tokio_core::reactor::Handle) {
+        match self {
+            Reactor::BlockReporter {
+                block_store,
+                validator_address,
+            } => {
+                auxiliary::report_block(event_loop, block_store, validator_address.clone(), block);
+            }
+        }
+    }
+}
+
+impl Reactor {
+    /// Register different kind of reactors to origin and auxiliary block chain.
+    ///
+    /// # Arguments
+    ///
+    /// * `origin` - A blockchain object that points to origin.
+    /// * `auxiliary` - A blockchain object that points to auxiliary.
+    /// * `config` - A configuration to register reactors.
+    pub fn register(origin: &mut Ethereum, auxiliary: &mut Ethereum, config: &Config) {
+        match origin.contract_instance(
+            config.origin_block_store_address(),
+            include_bytes!("./contract/abi/BlockStore.json"),
+        ) {
+            Ok(contract) => {
+                let origin_block_reporter = Reactor::BlockReporter {
+                    block_store: Arc::new(contract),
+                    validator_address: config.origin_validator_address(),
+                };
+                origin.register_reactor(origin_block_reporter);
+            }
+            Err(error) => {
```

if a function can fail it should return an error.